File: front/lib/api/src/models/artist.dart
import 'package:blee/api/src/models/image.dart';
import 'package:change_case/change_case.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'artist.freezed.dart';
part 'artist.g.dart';

@freezed
class Artist with _$Artist {
  const factory Artist({
    required String id,
    required String name,
    required String slug,
    required Image? poster,
  }) = _Artist;

  factory Artist.fromJson(Map<String, dynamic> json) => _$ArtistFromJson(json);
}

enum ArtistSort {
  name,
  addDate;

  @override
  String toString() {
    return this.name.toSnakeCase();
  }
}

File: front/models/domain/artist.ts
import * as yup from "yup";
import { Resource } from "./resource";
import { Image } from "./image";

const Artist = Resource.concat(
    yup.object({
        name: yup.string().required(),
        slug: yup.string().required(),
        poster: Image.nullable(),
    }),
);

type Artist = yup.InferType<typeof Artist>;
type ArtistSortingKeys = "name" | "addDate";

export { Artist, type ArtistSortingKeys };
